Added the Sponsorship model

narayanaditya95 opened the pull request:
@gvwilson @pbanaszkiewicz 
Writing tests. Ready for review.

narayanaditya95 commented on workshops/models.py:
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
+class Sponsorship(models.Model):
+    '''Represent sponsorship from a host for an event.'''
+
+    SPONSORSHIP_SLABS = (
```
@gvwilson @pbanaszkiewicz I would like to know whether a choice field would be better suited for SWC/DC workshops or a plain simple `IntegerField`?
If we go with choices, what are the slabs in which SWC/DC accept sponsorship.

pbanaszkiewicz replied:
For me slab is [this](http://peppardbuildingsupplies.com/sites/default/files/SWATCH_BS_Grey_0.jpg) :-)

Anyway, what does `Sponsorship.slab` represent here?

narayanaditya95 replied:
I am not sure how I ended up with such a field name.
Anyway, a `slab` is intended to show the divisions in which SWC/DC accept sponsorships. Each slab would have predefined sponsorship amount.
PyData uses it and it's sponsors can opt for divisions such as `PyData Founding`, `PyData Gold`, `PyData Silver`... each with decreasing sponsor amount and benefits.

narayanaditya95 commented on workshops/static/css/amy.css:
```diff
@@ -93,10 +93,9 @@ body {
 /* when using nav-tab (http://getbootstrap.com/javascript/#tabs) move the
    crispy form down a few notches */
-div[role="tabpanel"] form {
+.tab-content {
```
This CSS selector is more specific.
`div[role="tabpanel"] form` also added padding above the `Delete` buttons in the sponsorship table.

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
+    )
+
+    organization = models.ForeignKey(
+        'Host',
```
Switch to `Organization` after #849 is merged.

narayanaditya95 replied:
Yes.

pbanaszkiewicz replied:
That was just a note for future :-)

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -726,6 +755,10 @@ class Event(AssignmentMixin, models.Model):
         help_text='Organization responsible for administrative work.'
     )
+    sponsors = models.ManyToManyField(
+        Host, related_name='sponsor', blank=True,
```
Am I reading this correctly? `Event.sponsors → Hosts` but `Host.sponsor → Event`?

narayanaditya95 replied:
Yes, just as `host.administrator` gives the list of events that the host administered, `host.sponsor` gives the list of events sponsored.
On a second thought, I think `Event.sponsor_set` sounds better.

pbanaszkiewicz replied:
Ugh, both `organization.administrator` and `organization.sponsor` sound bad, and I believe I'm responsible for the former :-)

I'd use `organization.administered_events` or `organization.administered_set` and `organization.sponsored_events` or `organization.sponsored_set`.

pbanaszkiewicz commented on workshops/views.py:
```diff
@@ -156,6 +157,22 @@ def get_success_message(self, cleaned_data):
+        )
+        return super().delete(request, *args, **kwargs)
+
+
```
You sure you need this? We used to redirect right after deletion.

narayanaditya95 replied:
Yes, we currently also redirect right after deletion. There is no intermediate confirmation page (except for the confirm javascript popup).

pbanaszkiewicz replied:
So why do we need this view?

narayanaditya95 replied:
Because it makes the task of deleting and adding a success message easier and non-repetitive.

pbanaszkiewicz replied:
That's a nice change, can you do that for other deletion views if feasible (can be in a different PR)?

pbanaszkiewicz commented:
@narayanaditya95 looks OK.

I wonder what we should use for the sponsorship levels aka. slabs. You can imagine we don't have an established values for them in SWC/DC. Maybe @gvwilson can provide some more feedback here?

narayanaditya95 commented:
If SWC/DC does not use a predefined levels, we can switch to an `IntegerField`.
PyData package would then monkeypatch the `slab` field to add `choices` to it. Each predefined level at PyData has a sponsorship amount against it.

narayanaditya95 commented on workshops/models.py:
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
+    level = models.PositiveIntegerField(
+        choices=SPONSORSHIP_LEVELS,
+        default=1000,
+    )
```
@gvwilson Does SWC/DC use predefined levels of sponsorship, or should we simply remove the choices?

gvwilson replied:
Our current levels are defined on 
http://software-carpentry.org/scf/join/.  I think a Partnership model 
with names is overkill - just a text field naming the level should be 
enough, and we can add special notes to the Notes field on the 
organization where needed.

narayanaditya95 replied:
The levels defined at http://software-carpentry.org/scf/join/ list the kinds of partnership between an organization and SWC/DC. It does not list the kind of sponsorship an organization can provide for an SWC/DC workshop.
What practice does SWC/DC follow while accepting sponsorship for a particular workshop?

gvwilson replied:
My apologies, I misunderstood the question - organizational sponsorship 
for a specific workshop varies so widely that I don't think it's worth 
trying to capture as anything except a free-text field (in Notes).

narayanaditya95 commented:
@gvwilson I added a nullable `DecimalField` that would allow SWC/DC admins to record monetary sponsorship. The `pydata` app would on the other hand, extend the `DecimalField` to add predefined levels.

@pbanaszkiewicz I have added tests and addressed all the comments raised on the PR.
It is up for review and merge.

narayanaditya95 commented:
bump.

pbanaszkiewicz commented on workshops/templates/workshops/event_edit_form.html:
```diff
@@ -4,32 +4,68 @@
+  <ul class="nav nav-tabs" role="tablist">
+    <li role="presentation" class="active"><a href="#event" aria-controls="event" role="tab" data-toggle="tab">Event</a></li>
+    <li role="presentation"><a href="#tasks" aria-controls="tasks" role="tab" data-toggle="tab">Tasks</a></li>
+    <li role="presentation"><a href="#sponsors" aria-controls="sponsors" role="tab" data-toggle="tab">Sponsors</a></li>
```
See how `person-edit` page behaves after you add a new task to the person -- it switches you to a tab you just were at.

You'll receive 1000 bonus virtual points if you make the tabs switch depending on the anchor in the browser's address bar. :-)

pbanaszkiewicz replied:
PS the trick is to use additional tags with the `message.success`. See `person-edit` for details.

pbanaszkiewicz commented on workshops/views.py:
```diff
@@ -1194,7 +1214,20 @@ def event_edit(request, event_ident):
+
+                messages.success(
+                    request,
+                    '{} was added a new sponsor.'.format(sponsor.organization)
```
Should there be "as" instead of "a"?

pbanaszkiewicz commented on workshops/views.py:
```diff
@@ -1194,7 +1214,20 @@ def event_edit(request, event_ident):
+                    '{} was added a new sponsor.'.format(sponsor.organization)
+                )
+                # to reset the form values
+                return redirect(request.path)
```
…or `redirect(event)`, right? The latter is less popular but IMO a little bit clearer where you end up after the redirect.

pbanaszkiewicz commented:
@narayanaditya95 minor issues and one/two challenges, and apart from them +1 (again, not going to merge until after #873 is fixed).

narayanaditya95 commented:
@pbanaszkiewicz Where do I claim my 1000 bonus virtual points?

pbanaszkiewicz commented:
@narayanaditya95 good job, I'll merge soon into v1.7.

pbanaszkiewicz commented:
@narayanaditya95 can you please rebase? Thanks!

narayanaditya95 commented:
Done
